### PR TITLE
optimize single input column multi-value expressions

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR follows up on the changes in #8014, which actually opened the door to using the same optimized selector to evaluate single input column expressions for multi-value string columns, using native Druid behavior to do the `map` of the expression across the values. This works provided a few conditions are met: 
* the expression _does not_ operate on the multi-value input as an array
* the expression _does not_ produce an array
* the single multi-value column is used _at most once_ in the expression

To achieve this, functions and apply functions now also report if they have array inputs or outputs which is collected on `Expr.BindingDetails`, which is already used to analyze the expression in `ExpressionSelectors` in order to determine the correct type of selector to create.

This change also allows multi-value expressions that meet the above conditions to work with group by v1, since the cardinality can be known, however wider multi-value expression support for group by v1 is not available (or really my priority).

This PR has also uncovered some inconsistencies in the null value handling for multi-value expressions, [specifically the part that coerces `null`, `[]`, and `[null]` all into `[null]`](https://github.com/apache/incubator-druid/blob/master/processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java#L446), which creates some minor differences between using the `SingleStringInputDimensionSelector` to "map" an expression across the row values compared to using a `map` expression. Ideally I will modify the coercion to not convert `[]` to `[null]`, which seems to be the problematic issue, but it creates mass of failing tests in the `ExpressionFilterTest` due to differences in behavior in the underlying selectors, so I intend to investigate this deeper in follow-up work The difference in behavior is apparent in the modified calcite test, where `[]` is grouped as null with default null handling behavior instead of `""`. This is likely related or at least similar to #5897 and #4915.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths.

### CODE REVIEW

**Comment by clintropolis** [comment]
Tagged for milestone 0.16.0 since this should be included in the same release with the rest of the multi-value expression support, to avoid a future behavior change on single input multi-value expressions.

File: core/src/main/java/org/apache/druid/math/expr/Expr.java [hunk 1]
```diff
@@ -248,6 +256,24 @@ public BindingDetails(
       return arrayVariables.stream().map(IdentifierExpr::getIdentifier).collect(Collectors.toSet());
     }
 
+    /**
+     * Returns true if the expression has any array inputs. Note that in some cases, this can be true and
+     * {@link BindingDetails#getArrayVariables} can be empty, since the latter can only shallowly collect identifiers
+     * that are explicitly used as arrays.
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/math/expr/Expr.java]: I turned this comment over a few times in my mind but don't quite understand it. What does "shallowly" mean in this context and what is "deeper" about `hasInputArrays`?

File: core/src/main/java/org/apache/druid/math/expr/Expr.java [hunk 2]
```diff
@@ -304,7 +334,37 @@ public BindingDetails withArrayArguments(Set<Expr> arrayArguments)
       return new BindingDetails(
           ImmutableSet.copyOf(Sets.union(freeVariables, arrayIdentifiers)),
           scalarVariables,
-          ImmutableSet.copyOf(Sets.union(arrayVariables, arrayIdentifiers))
+          ImmutableSet.copyOf(Sets.union(arrayVariables, arrayIdentifiers)),
+          hasInputArrays || arrayArguments.size() > 0,
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/math/expr/Expr.java]: Is it possible for `arrayArguments.size()` to be greater than zero, but `hasInputArrays` to be false? If so how? (Related to previous comment, I think)

File: core/src/main/java/org/apache/druid/math/expr/Expr.java [hunk 3]
```diff
@@ -824,7 +888,12 @@ public BindingDetails analyzeInputs()
     }
 
     lambdaBindingDetails = lambdaExpr.analyzeInputs();
-    bindingDetails = accumulator.with(lambdaBindingDetails).withArrayArguments(function.getArrayInputs(argsExpr));
+    boolean isArrayOutput = function.hasArrayOutput() || (function instanceof ApplyFunction.BaseFoldFunction
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/math/expr/Expr.java]: Using `instanceof` here seems like a code smell -- is there a nice way to add a method to `ApplyFunction` that gets the necessary info out?

File: core/src/main/java/org/apache/druid/math/expr/Function.java [hunk 4]
```diff
@@ -76,6 +76,22 @@
     return Collections.emptySet();
   }
 
+  /**
+   * Returns true if a function takes an array arguments
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/math/expr/Function.java]: 1. "takes array arguments" or "takes an array argument" (grammar, and also, I'm not sure which one you mean; the two phrases mean different things)
2. Clarify intent: what does "takes an array arguments" mean? That it _might_ take array arguments? Or that it _must_ take array arguments?

File: core/src/main/java/org/apache/druid/math/expr/Function.java [hunk 5]
```diff
@@ -76,6 +76,22 @@
...
+   */
+  default boolean hasArrayInputs()
+  {
+    return false;
+  }
+
+  /**
+   * Returns true if function produces an array
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/math/expr/Function.java]: What happens if it sometimes produces an array, and sometimes doesn't, based on its input type? Similar comment to `ApplyFunction.hasArrayOutput`: is it disallowed, or should it return true/false?

File: processing/src/test/java/org/apache/druid/segment/filter/ExpressionFilterTest.java [hunk 6]
```diff
@@ -147,6 +147,7 @@ public void testOneMultiValuedStringColumn()
     }
     assertFilterMatchesSkipVectorize(edf("dim4 == '1'"), ImmutableList.of("0"));
     assertFilterMatchesSkipVectorize(edf("dim4 == '3'"), ImmutableList.of("3"));
+    assertFilterMatchesSkipVectorize(edf("dim4 == '4'"), ImmutableList.of("4", "5"));
```
Comment (gianm) [inline processing/src/test/java/org/apache/druid/segment/filter/ExpressionFilterTest.java]: Why'd this change?

File: processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java [hunk 7]
```diff
@@ -160,7 +160,7 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
                  && capabilities.isDictionaryEncoded()
                  && capabilities.isComplete()
                  && !capabilities.hasMultipleValues()
-                 && !exprDetails.getArrayColumns().contains(column)) {
+                 && exprDetails.getArrayColumns().size() == 0) {
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java]: `isEmpty()`?

File: processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java [hunk 8]
```diff
@@ -230,8 +230,9 @@ public static DimensionSelector makeDimensionSelector(
           && capabilities.getType() == ValueType.STRING
           && capabilities.isDictionaryEncoded()
           && capabilities.isComplete()
-          && !capabilities.hasMultipleValues()
-          && !exprDetails.getArrayColumns().contains(column)
+          && !exprDetails.hasInputArrays()
+          && !exprDetails.isOutputArray()
+          && (!capabilities.hasMultipleValues() || exprDetails.getFreeVariables().size() == 1)
       ) {
         // Optimization for dimension selectors that wrap a single underlying string column.
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java]: The flag chain above is getting complicated!

Could you please alter this comment to reflect what it's looking for? If I understand it right, that should be something like:

> Optimization for dimension selectors that wrap a single underlying string column. The string column can be multi-valued, but if so, it must be implicitly mappable (i.e. the expression is not treating it as an array, not wanting to output an array, and the multi-value dimension appears exactly once).

By the way, I have to assume logic like this is being used elsewhere to implicitly map expressions that go through the non-optimized selector. Is there any opportunity to extract the logic into a common helper method? It would make it easier for readers and probably less error prone too.

File: core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java [hunk 9]
```diff
@@ -60,6 +60,17 @@
    */
   Set<Expr> getArrayInputs(List<Expr> args);
 
+  /**
+   * Returns true if apply function produces an array output
+   */
+  default boolean hasArrayOutput()
+  {
+    return false;
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java]: This should not be allowed imo, but I think we'll have to discuss this further when we get around to adding full type inference to the expression system. I will adjust the javadoc to state it should not be done currently.

File: core/src/main/java/org/apache/druid/math/expr/Expr.java (hunk 1 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Expr.java]: The distinction is that `getArrayVariables` returns the identifiers we were able to explicitly detect as arrays, where as `hasInputArrays` means that an expression expects some number of array inputs. `Expr.analyzeInputs` implementations can currently only correctly categorize arguments that are _directly_ identifiers. 

Now, the reason we need to be able to distinguish is an issue because I also accept _slop_ expressions. For example:

```
array_to_string(concat('foo', tags), ', ')
``` 

where `tags` is a multi-value dimension, the expression bindings will list `tags` as a scalar input, but `array_to_string` _does_ in fact take an array argument. By distinguishing these I can still accept this expression and translate it correctly to:

```
array_to_string(map((tags) -> concat('foo', tags), tags), ',')
```
and not _incorrectly_ use this optimized single input column introduced in this PR. This was not an issue prior to this PR because being multi-valued in the capabilities was reason enough to exclude an expression from the single column optimization, but after the restriction only applies to things that expects array inputs or produce array outputs. 

Does that make sense? I will try to clear up the javadocs for this.

File: core/src/main/java/org/apache/druid/math/expr/Expr.java (hunk 2 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Expr.java]: I don't think so, this can probably just check `hasInputArrays`, but I will confirm by removing the size check and seeing what happens to the tests

File: core/src/main/java/org/apache/druid/math/expr/Expr.java (hunk 3 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Expr.java]: Maybe? Will have a look

File: core/src/main/java/org/apache/druid/math/expr/Function.java (hunk 4 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Function.java]: the typo is that it's supposed to be `any array arguments` heh, which I think clears up both things if I fix?

File: core/src/main/java/org/apache/druid/math/expr/Function.java (hunk 5 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Function.java]: see above comment, will adjust docs to mention it shouldn't be done, but i don't have a way to really enforce it yet, need type inference i think

File: processing/src/test/java/org/apache/druid/segment/filter/ExpressionFilterTest.java (hunk 6 above)
Comment (clintropolis) [reply processing/src/test/java/org/apache/druid/segment/filter/ExpressionFilterTest.java]: I .. don't really remember, I'm going to go with the reason was for fun, to add another test that actually has multiple matches in the case where sql compatible null handling is enabled 😜

File: processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java (hunk 8 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java]: The logic is only used in these 2 places in ExpressionSelectors I think, and it's slightly different between the uses because `SingleStringInputDimensionSelector` supports multi-value inputs, but `SingleStringInputCachingExpressionColumnValueSelector` does not. I'll play around and see what I can do, or at least comment what's going on heavily.

File: core/src/main/java/org/apache/druid/math/expr/Expr.java (hunk 2 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Expr.java]: Oh i wasn't paying attention to which function this was, it is possible in this specific context and it's unrelated to the other thing. 

Since this is a sort of 'builder' method it is hypothetically possible for this method to be called when adding the initial set of array columns on a binding details, without explicitly calling `withArrayInputs`. This `||` then ensures that if even if you forgot to call `withArrayInputs` the binding details being constructed are correct, becase the `hasInputArrays` in this case is the value of the existing binding before the array arguments are added.

File: core/src/main/java/org/apache/druid/math/expr/Expr.java (hunk 3 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/Expr.java]: I've modified `ApplyFunction.hasArrayOutput` to now take the `LambaExpr` as an argument so `BaseFoldFunction` can just override `hasArrayOutput` to examine the lambda expression to see if it's output is array typed.

File: core/src/main/java/org/apache/druid/math/expr/Function.java [hunk 10]
```diff
@@ -1801,24 +1830,30 @@ static void setArrayOutputElement(
...
-      }
+      return true;
     }
 
     @Override
-    public Set<Expr> getScalarInputs(List<Expr> args)
+    public void validateArguments(List<Expr> args)
     {
-      return ImmutableSet.copyOf(args);
+      if (!(args.size() > 0)) {
```
Comment (ccaominh) [inline core/src/main/java/org/apache/druid/math/expr/Function.java]: Could simplify to:
```java
if (args.isEmpty()) {
```
